src/crypto/openssl/key_expansion.rs

```diff
@@ -5,46 +5,87 @@ use crate::{
     crypto::{
         cipher_suite::{CipherSuite, CipherSuiteVariant},
         key_expansion::{
-            ExpandAsSecret, KeyMaterial, SFRAME_HDKF_SALT_EXPAND_INFO, SFRAME_HKDF_KEY_EXPAND_INFO,
-            SFRAME_HKDF_SALT,
+            KeyExpansion, SFRAME_HDKF_SALT_EXPAND_INFO, SFRAME_HDKF_SUB_AUTH_EXPAND_INFO,
+            SFRAME_HKDF_KEY_EXPAND_INFO, SFRAME_HKDF_SALT, SFRAME_HKDF_SUB_ENC_EXPAND_INFO,
+            SFRAME_HKDF_SUB_SALT,
         },
         secret::Secret,
     },
     error::{Result, SframeError},
 };
 
-impl ExpandAsSecret for KeyMaterial<'_> {
-    fn expand_as_secret(&self, cipher_suite: &CipherSuite) -> Result<Secret> {
+impl KeyExpansion for Secret {
+    fn expand_from<T>(cipher_suite: &CipherSuite, key_material: T) -> Result<Secret>
+    where
+        T: AsRef<[u8]>,
+    {
         let try_expand = || {
-            let prk = extract_prk(cipher_suite, self.0)?;
-            let key = expand_key(
-                cipher_suite,
-                &prk,
-                SFRAME_HKDF_KEY_EXPAND_INFO,
-                cipher_suite.key_len,
-            )?;
-            let salt = expand_key(
-                cipher_suite,
-                &prk,
-                SFRAME_HDKF_SALT_EXPAND_INFO,
-                cipher_suite.nonce_len,
-            )?;
-
-            Ok(Secret { key, salt })
+            let (base_key, salt) = expand_secret(cipher_suite, key_material.as_ref())?;
+            let (key, auth) = if cipher_suite.is_ctr_mode() {
+                let (key, auth) = expand_subsecret(cipher_suite, &base_key)?;
+                (key, Some(auth))
+            } else {
+                (base_key, None)
+            };
+
+            Ok(Secret { key, salt, auth })
         };
 
         try_expand().map_err(|_: openssl::error::ErrorStack| SframeError::KeyExpansion)
```

Comment:
this eats errors, maybe we want to at least log the stack

Reply:
Will add a log

( looking e.g. at `sender.rs` we should anyhow iterate over our logs again ;) )